File: graphs/ltcd_765.py
'''
LeetCode - problem 765
'''

    def minSwapsCouples(self, row):
        x=row
        indecies={x[i]:i for i in range(len(x))}
        swap=0

        for i in range (0,len(x),2):
            a=x[i]
            b=x[i+1]
            if ((a%2==0 and b==a+1)or (b%2==0 and a==b+1)):
                continue
            if a%2==0:
                target=a+1
            else:
                target=a-1
            target_index=indecies[target]
            b_index=i+1

            x[target_index],x[b_index]= x[b_index],x[target_index]
            indecies[target]=b_index
            indecies[b]=target_index

            swap +=1
        return (swap)

File: graphs/ltcd_765.cpp
//LeetCode - problem 765

class Solution {
    vector<int> parent;
public:

    int find(int x){
        if(parent[x]==x){
            return x;
        }
        parent[x] = find(parent[x]);
        return parent[x];
    }
    
    bool Union(int x, int y){
        int a = find(x);
        int b = find(y);
        if(a==b)return false;
        parent[a] = b;
        return true;
    }
    
    int minSwapsCouples(vector<int>& row) {
        parent.resize(row.size());
        
        for(int i=0;i<row.size();i+=2){
            parent[row[i]] = row[i];
            parent[row[i+1]] = row[i];
        }
        
        int counter = 0;
        for(int i=0;i<row.size();i+=2){
            if(Union(i,i+1)){
                counter++;
            }
        }
       return counter; 
    }
};
